File: 19-08-2025_Farthest_Smaller_Right.java
import java.util.*;

class Solution {
    public int[] farMin(int[] arr) {
        int n = arr.length;
        
        int[] suffix = new int[n + 1];
        Arrays.fill(suffix, Integer.MAX_VALUE);

        for (int i = n - 1; i >= 0; i--) {
            suffix[i] = Math.min(suffix[i + 1], arr[i]);
        }

        int[] ans = new int[n];
        Arrays.fill(ans, -1);

        for (int i = 0; i < n - 1; i++) {
            int low = i + 1;
            int high = n - 1;
            int res = -1;

            while (low <= high) {
                int mid = (low + high) / 2;

                if (suffix[mid] < arr[i]) {
                    res = mid;
                    low = mid + 1; // move right to find farthest
                } else {
                    high = mid - 1;
                }
            }
            ans[i] = res;
        }
        return ans;
    }
}

File: 19-08-2025_Farthest_Smaller_Right.py
from typing import List

class Solution:
    def farMin(self, arr: List[int]) -> List[int]:
        n = len(arr)
        
        suffix = [float('inf')] * (n + 1)
        for i in range(n - 1, -1, -1):
            suffix[i] = min(suffix[i + 1], arr[i])
        
        ans = [-1] * n
        for i in range(n - 1):
            low, high = i + 1, n - 1
            res = -1
            while low <= high:
                mid = (low + high) // 2
                if suffix[mid] < arr[i]:
                    res = mid
                    low = mid + 1  # go right
                else:
                    high = mid - 1
            ans[i] = res
        return ans
